[UI Framework] add KuiExpressionItem and related components.

PopradiArpad opened the pull request:
Implement

KuiExpressionItem
KuiExpressionItemButton
KuiExpressionItemPopover
and theirs test.
Update docs examples.

PopradiArpad commented on ui_framework/components/expression/expressionItemPopover.js:
```diff
@@ -0,0 +1,98 @@
+https://stackoverflow.com/questions/32553158/detect-click-outside-react-component
+*/
+const outsideClickNotifier = WrappedComponent => class extends React.Component {
+  constructor(props) {
```
I wanted to separate the stateless from the statefull.
  Stateless is KuiExpressionItemPopover, statefull is outsideClickNotifier.
  Probably let outsideClickNotifier reuse. If it's not worth to do it
  I would merge it into KuiExpressionItemPopover.

PopradiArpad commented on ui_framework/components/expression/expressionItemPopover.js:
```diff
@@ -0,0 +1,98 @@
+  children: PropTypes.node,
+  align: PropTypes.oneOf(POPOVER_ALIGN),
+  //only for outsideClickNotifier
+  rootRef: PropTypes.func.isRequired,
```
A drawback that rootRef prop is visible on the interface and can cause confusion.

PopradiArpad commented:
I add tests after the interface is fixed.

cjcenizal commented:
Awesome!! @PopradiArpad this week has been crazy for me, and I've been meaning to review this PR as soon as possible. I think I'll get to it next week. Thanks so much for doing this.

cjcenizal commented:
@PopradiArpad The structure of files in the `ui_framework` directory has changed. We just moved the `components` directory to `src/components`. Would you mind rebasing to bring this branch up to date with master? To resolve the merge conflicts you'll need to do 3 things:

1. Move the files inside `components/expression` into `src/components/expression`.
2. Make sure that `components/index.js` consists solely of: `export * from '../src/components';`
3. Update `src/components/index.js` with your exports:
```javascript
export {
  KuiExpressionItem,
  KuiExpressionItemButton,
  KuiExpressionItemPopover,
} from './expression';
```

PopradiArpad commented:
@cjcenizal No problem with the late answer. Me too have to do other things as well.
Is the interface of the expression components right? I wanted to stay at no abstraction to let experiment it and abstract only if more uses cases are defined. 
But outsideClickNotifier is for component intern usage. Such capability could be useful for other compenents too. What do you think about it in KuiExpressionItemPopover?

cjcenizal commented on ui_framework/src/components/expression/expressionItemPopover.js:
```diff
@@ -0,0 +1,98 @@
+  isVisible,
+  children,
+  rootRef,
+  align = 'left',
```
Can we use `defaultProps` to set this default value?

PopradiArpad replied:
ok

cjcenizal commented:
Can you also add `hasReact` for this component in `routes.js`?

PopradiArpad commented on ui_framework/src/components/expression/expressionItem.js:
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
```
oops...

PopradiArpad commented on ui_framework/src/components/expression/expressionItemButton.js:
```diff
@@ -0,0 +1,35 @@
+import React from 'react';
```
...hmmm

PopradiArpad commented on ui_framework/src/components/expression/expressionItemButton.js:
```diff
@@ -0,0 +1,35 @@
+      {...rest}
+    >
+      <span className="kuiExpressionItem__buttonDescription">{description}</span>{' '}
+      <span className="kuiExpression_buttonValue">{value}</span>
```
ok.

PopradiArpad commented on ui_framework/src/components/expression/expressionItemButton.js:
```diff
@@ -0,0 +1,35 @@
+KuiExpressionItemButton.propTypes = {
+  className: PropTypes.string,
+  description: PropTypes.string.isRequired,
+  value: PropTypes.string.isRequired,
```
good idea

PopradiArpad commented on ui_framework/src/components/expression/expressionItemPopover.js:
```diff
@@ -0,0 +1,98 @@
+import React from 'react';
```
of course :)

PopradiArpad commented:
hasReact added too. I make the tests.

PopradiArpad commented:
@cjcenizal I don't test the onOutsideClick property on KuiExpressionItemPopover because as I see it can be tested right only in a headless browser.

PopradiArpad commented:
I adapted KuiExpressionItemPopover to a local version of your KuiOutsideClickDetector.
So KuiExpressionItemPopover got rid of the ugly rootRef property.
In example_new I let show the popover only if a button is clicked which spares a lot of complications
with the event handling (races between click events and worrying about invisible things).
I haven't replaced the example yet to let play and see both the CSS and the React component code.

cjcenizal commented:
@PopradiArpad OK, I merged the `KuiOutsideClickDetector` PR! Would you mind rebasing master and updating this component to use it?

PopradiArpad commented:
@cjcenizal I rebased,  `KuiOutsideClickDetector` is in use and I removed the old example.

zinckiwi commented on ui_framework/doc_site/src/views/expression/expression.js:
```diff
@@ -1,26 +1,139 @@
+    this.changeExample2Object = this.changeExample2Object.bind(this);
+    this.changeExample2Description = this.changeExample2Description.bind(this);
+    this.changeExample2Value = this.changeExample2Value.bind(this);
+    this.onOutsideClick = this.onOutsideClick.bind(this);
```
The following style avoids having to manually bind in this way:

```
changeExample1: event => {
  this.setState(_.merge(this.state,{ example1:{ value: event.target.value } }));
}
```

PopradiArpad replied:
@zinckiwi Class properties are cool!

weltenwort commented on ui_framework/src/components/expression/expression_item_popover.js:
```diff
@@ -0,0 +1,57 @@
+  children: PropTypes.node,
+  align: PropTypes.oneOf(POPOVER_ALIGN),
+  onOutsideClick: PropTypes.func.isRequired,
+};
```
What do you think about not requiring `isVisible` and `onOutsideClick` and providing `defaultProps` for them (and `align: 'left'`)? That would make the tests a bit less verbose.

PopradiArpad replied:
@weltenwort Thank you for that!
`isVisible` is not needed anymore because `KuiExpressionItemPopover`
will be added to `KuiExpressionItem` only if the popover is needed.
I remove `isVisible` and `.kuiExpressionItem__popover--isHidden`
and adapt the tests.

I assume the default usage of `KuiExpressionItemPopover`
is to vanish for an outside click and to accomplish that the client must be
notified about the outside click, so as I see it's better to let `onOutsideClick` explicit.

`align: 'left' `has already a default prop.

weltenwort commented on ui_framework/src/components/expression/expression_item_popover.test.js:
```diff
@@ -0,0 +1,100 @@
+  test('renders', () => {
+    const component = (
+      <KuiExpressionItemPopover
+        title={'title'}
```
This works, but the canonical form would be `title="title"`. The same applies to some of the other props below.

PopradiArpad replied:
@weltenwort OK. I hope I changed that in all files.

cjcenizal commented:
FYI we've made a few additions to our linting rules, and you might need to reinstall your node dependencies to get them.

cjcenizal commented:
Also just a note that this will need to be backported to 6.x.

PopradiArpad commented on ui_framework/doc_site/src/views/expression/expression.js:
```diff
@@ -1,26 +1,139 @@
+  }
+
+  changeExample1(event) {
+    this.setState(_.merge(this.state,{ example1:{ value: event.target.value } }));
```
@weltenwort Oops! I overlooked the spaces.
The object spread operator is great!

PopradiArpad commented on ui_framework/doc_site/src/views/expression/expression.js:
```diff
@@ -1,26 +1,139 @@
+        onOutsideClick={this.onOutsideClick}
+        style={popoverStyle}
+      >
+        <select className="kuiSelect" value={this.state.example1.value} onChange={this.changeExample1}>
```
@weltenwort You are right. I change them.

PopradiArpad commented on ui_framework/doc_site/src/views/expression/expression_example.js:
```diff
@@ -8,30 +9,30 @@ import {
-const expressionJs = require('raw!./expression.js');
+const Expression = require('./expression');
+const expressionSource = require('!!raw!./expression');
+const expressionHtml = renderToHtml(Expression);
```
@weltenwort Good idea! 
Showing an expression popup by default can be a use case. I do it.

PopradiArpad commented:
@cjcenizal @weltenwort @zinckiwi
I rebased on top of the newest master and reinstalled the project.
I removed the isVisible property from KuiExpressionItemPopover and
the .kuiExpressionItem__popover--isHidden CSS class.
The tests and the example are adapted.
I removed expression.html from the examples.

weltenwort commented on ui_framework/src/components/expression/_expression.scss:
```diff
@@ -48,12 +48,6 @@
 
     // 1. Angulars ng-hide uses display: none. To use animations we need to use visibility instead.
     &.ng-hide,
```
Now `&.ng-hide` shares the block with `&:before`, which is probably not intended. I think the previous block should remain for `&.ng-hide` to avoid a flash of unstyled content.

PopradiArpad replied:
@weltenwort Ok.

weltenwort commented:
jenkins, test this

PopradiArpad commented:
@weltenwort Is the pull request ok now?

weltenwort commented:
jenkins, test this

weltenwort commented:
:tada: thanks so much @PopradiArpad

PopradiArpad commented:
@cjcenizal @weltenwort @zinckiwi Thank you for all the great suggestions!!!

cjcenizal commented:
Thanks for your hard work @PopradiArpad!
